File: modules/text/samples/dictnet_demo.cpp
#include  "opencv2/text.hpp"
#include  "opencv2/highgui.hpp"
#include  "opencv2/imgproc.hpp"

#include  <sstream>
#include  <iostream>

using namespace std;
using namespace cv;
using namespace cv::text;

inline void printHelp()
{
    cout << "    Demo of wordspotting CNN for text recognition." << endl;
    cout << "    Max Jaderberg et al.: Reading Text in the Wild with Convolutional Neural Networks, IJCV 2015"<<std::endl<<std::endl;

    cout << "    Usage: program <input_image>" << endl;
    cout << "    Caffe Model files  (dictnet_vgg.caffemodel, dictnet_vgg_deploy.prototxt, dictnet_vgg_labels.txt)"<<endl;
    cout << "      must be in the current directory." << endl << endl;

    cout << "    Obtaining Caffe Model files in linux shell:"<<endl;
    cout << "    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg.caffemodel"<<endl;
    cout << "    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg_deploy.prototxt"<<endl;
    cout << "    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg_labels.txt"<<endl<<endl;
}

int main(int argc, const char * argv[])
{
    if (argc != 2)
    {
        printHelp();
        exit(1);
    }

    Mat image = imread(argv[1], IMREAD_GRAYSCALE);

    cout << "Read image (" << argv[1] << "): " << image.size << ", channels: " << image.channels() << ", depth: " << image.depth() << endl;

    if (image.empty())
    {
        printHelp();
        exit(1);
    }

    Ptr<OCRHolisticWordRecognizer> wordSpotter = OCRHolisticWordRecognizer::create("dictnet_vgg_deploy.prototxt", "dictnet_vgg.caffemodel", "dictnet_vgg_labels.txt");

    std::string word;
    vector<float> confs;
    wordSpotter->run(image, word, 0, 0, &confs);

    cout << "Detected word: '" << word << "', confidence: " << confs[0] << endl;
}

File: modules/text/samples/dictnet_demo.py
#!/usr/bin/env python

import cv2
import sys
import os.path
 
#Global variable shared between the Mouse callback and main
refPt = []
cropping = False
image=None
drawImage=None
dictNet=None


def mouseCallback(event, x, y, flags, param):
    # grab references to the global variables
    global refPt, cropping,dictNet,drawImage,image

    # if the left mouse button was clicked, record the starting
    # (x, y) coordinates and indicate that cropping is being
    # performed
    if event == cv2.EVENT_LBUTTONDOWN:
        refPt = [(x, y)]
        cropping = True

    # check to see if the left mouse button was released
    elif event == cv2.EVENT_LBUTTONUP:
        # record the ending (x, y) coordinates and indicate that
        # the cropping operation is finished
        refPt.append((x, y))
        cropping = False

        # draw a rectangle around the region of interest
        roi = image[refPt[0][1]:refPt[1][1], refPt[0][0]:refPt[1][0]]
        res=dictNet.classify([roi])
        drawImage = image.copy()
        cv2.rectangle(drawImage, refPt[0], refPt[1], (0, 255, 0), 2)
        cv2.putText(drawImage,"%s:%f"%(res[0][0],res[1][0]),refPt[0],cv2.FONT_HERSHEY_SIMPLEX,1,(255,255,255),2)
        cv2.imshow("Select A Region", drawImage)
        

if __name__=='__main__':
    helpStr="""    Usage: """+sys.argv[0]+""" IMAGE_FILENAME
    
    Press 'q' or 'Q' exit
    
    The modelFiles must be available in the current directory.
    In linux shell they can be downloaded (~2GB) with the following commands:
    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg.caffemodel
    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg_deploy.prototxt
    wget http://nicolaou.homouniversalis.org/assets/vgg_text/dictnet_vgg_labels.txt
    """
    if((len(sys.argv)!=2 )or not(os.path.isfile(sys.argv[1]) )):
        print helpStr
        print 'No image file given Aborting!'
        sys.exit(1)
    if not (os.path.isfile('dictnet_vgg_deploy.prototxt') and 
        os.path.isfile('dictnet_vgg.caffemodel') and
        os.path.isfile('dictnet_vgg_labels.txt')):
        print helpStr
        print 'Model files not present, Aborting!'
        sys.exit(1)

    dictNet=cv2.text.OCRDictnet_create('dictnet_vgg_deploy.prototxt',
                                       'dictnet_vgg.caffemodel',
                                       'dictnet_vgg_labels.txt','',1,0,1)
    image = cv2.imread(sys.argv[1])
    drawImage = image.copy()
    cv2.namedWindow("Select A Region")
    cv2.setMouseCallback("Select A Region", mouseCallback)

    while True:
        cv2.imshow("Select A Region", drawImage)
        key = cv2.waitKey(1) & 0xFF

        # if the 'q' key is pressed, break from the loop
        if key == ord("q") or key == ord("Q"):
            break

    cv2.destroyAllWindows()